docs/integrations/custom_middleware.md

```diff
@@ -0,0 +1,191 @@
+```
+
+The first argument will be a Mirascope `CallResponse` of the provider you are using.
+The second argument will be your Mirascope call function.
+The third argument is the yielded object of your `custom_context_manager`
+
+`handle_call_response_async` is the same as `handle_call_response` but using an `async` function.
+
+### `handle_stream` and `handle_stream_async`
+
+These functions will be called after streaming a Mirascope call with the following signature:
+
+```python
+from typing import Callable
+
+from mirascope.core.base._stream import BaseStream
+
+
+def handle_stream(
+    stream: BaseStream, fn: Callable, context_manager: ContextManagerType | None
+):
+    # handle after stream here
+```
+
+The first argument will be a Mirascope `Stream` of the provider you are using.
+See `handle_call_response` section for information regarding the other arguments.
+
+One thing to note for `handle_stream` is that it will not be called until after the `Generator` has been exhausted.
+
+### `handle_response_model` and `handle_response_model_async`
+
+These functions will be called after making a mirascope call with `response_model` set with the following signature:
+
+```python
+from typing import Callable
+from pydantic import BaseModel
+
+from mirascope.core.base._utils._base_type import BaseType
+
+def handle_response_model(
+    response_model: BaseModel | BaseType, fn: Callable, context_manager: ContextManagerType | None
+):
+    # handle after call with response_model here
+```
+
+The first argument will be a Pydantic `BaseModel` or Python primative depending on what type `response_model` is.
+See `handle_call_response` section for information regarding the other arguments.
+
+For `BaseModel` you can grab the `CallResponse` via `response_model._response`.
+However for primatives `BaseType`, this information is not available.
+
+### `handle_structured_stream` and `handle_structured_stream_async`
+
+These functions will be called after streaming a mirascope call with `response_model` set with the following signature:
+
+```python
+from typing import Callable
+
+from mirascope.core.base._structured_stream import BaseStructuredStream
+
+def handle_structured_stream(
+    result: BaseStructuredStream,
+    fn: Callable,
+    context_manager: ContextManagerType | None,
+):
+```
+
+The first argument will be a Mirascope `StructuredStream` of the provider you are using.
+Like with `handle_stream`, `handle_structured_stream` will not be called until the `Generator` has been exhausted.
+
+### `custom_context_manager`
+
+You can define your own context manager where the yielded object is passed to each of the handlers.
+
+Here is an example using [SQLModel](https://sqlmodel.tiangolo.com/):
+
+```python
+from contextlib import contextmanager
+from typing import Any, Callable, Generator
+
+from sqlmodel import Session
+
+from mirascope.core.base import BaseCallResponse
+
+
+@contextmanager
+def custom_context_manager(
+    fn: Callable,
```

Comment:
confusing why we pass `fn` here at all, so I feel like we should either show using it or explain what it's for?

Reply:
Should I not show it at all? What we use it for is to grab function name, but not sure how it would be used for SQLModel

Reply:
can `custom_context_manager` be a callable that takes no arguments?

Reply:
I always pass in fn

Reply:
ok then we should at least better describe what `fn` is and where they might use it. If the answer is that we're only using it for the function name, then I wonder if we should instead just specifically provide the function name? What else would we need the function for here when we don't even really know what it is / how to call it and don't expect to (shouldn't?) call it.

Reply:
i dont think we should make any assumptions on what people might want to do. are you speaking from the perspective of a user? they know its a callable and I have already described what fn was above.